raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,172 @@
+import json
+
+from web3 import Web3
+
+from raiden.storage.sqlite import EventRecord, SQLiteStorage, StateChangeRecord
+from raiden.utils.typing import Any, BlockNumber, Dict
+
+SOURCE_VERSION = 18
+TARGET_VERSION = 19
+
+
+class BlockHashCache(object):
+    """A small cache for blocknumber to blockhashes to optimize this migration a bit
+
+    This cache lives only during the v18->v19 migration where numerous RPC calls are
+    expected to be made, many of them probably querying the same block_number -> block_hash.
+    To save RPC calls and make it a bit faster we keep this short-lived cache for the duration
+    of the migration
+    """
+    def __init__(self, web3: Web3):
+        self.web3 = web3
+        self.mapping = {}
+
+    def get(self, block_number: BlockNumber) -> str:
+        """Given a block number returns the hex representation of the blockhash"""
+        if block_number in self.mapping:
+            return self.mapping[block_number]
+
+        block_hash = self.web3.eth.getBlock(block_number)['hash']
+        block_hash = block_hash.hex()
+        self.mapping[block_number] = block_hash
+        return block_hash
+
+
+def _add_blockhash_to_state_changes(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to ContractReceiveXXX and ActionInitChain state changes"""
+    state_changes = storage.get_all_state_changes()
+    updated_state_changes = []
+    for state_change in state_changes:
+        data = json.loads(state_change.data)
+        affected_state_change = (
+            'raiden.transfer.state_change.ContractReceive' in data['_type'] or
+            'raiden.transfer.state_change.ActionInitChain' in data['_type']
+        )
+        if affected_state_change:
+            assert 'block_hash' not in data, 'v18 state changes cant contain blockhash'
+            block_number = int(data['block_number'])
+            data['block_hash'] = cache.get(block_number)
+
+            updated_state_changes.append(StateChangeRecord(
+                state_change_identifier=state_change.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+    storage.update_state_changes(updated_state_changes)
+
+
+def _add_blockhash_to_events(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to all ContractSendXXX events"""
+    events = storage.get_all_event_records()
+    updated_events = []
+    for event in events:
+        data = json.loads(event.data)
+        if 'raiden.transfer.events.ContractSend' in data['_type']:
```

Comment:
I would also reverse the condition here and dedent the rest of the code. Reason being less levels i have to pay attention to when reading the code.

Reply:
Sure good point.